File: 2022/day2.py
import input

TO_INT = {'X': 1, 'A': 1, "Y": 2, "B": 2, "Z": 3, "C": 3}
WIN = {1: 2, 2: 3, 3: 1}
LOSE = {2: 1, 3: 2, 1: 3}
    
def parse_data():
    data = input.retrieve(__file__)
    return [[TO_INT[move] for move in r.split(" ")] for r in data.split("\n")]

def points(them, me):
    if me == them:
        return 3 + me
    elif me == WIN[them]:
        return 6 + me
    else:
        return 0 + me
    
def play(them, outcome):
    if outcome == 2:
        return them
    elif outcome == 1:
        return LOSE[them]
    else:
        return WIN[them]
            
print(f"Part 1: {sum([points(t, m) for t, m in parse_data()])}")
print(f"Part 2: {sum([points(t, play(t, o)) for t, o in parse_data()])}")

File: 2022/rust/src/day2.rs
use std::collections::HashMap;

const INPUT: &str = include_str!("day2.txt");
const WIN: [i32; 3] = [2, 3, 1];
const LOSE: [i32; 3] = [3, 1, 2];

fn input() -> Vec<Vec<i32>> {
    let to_int: HashMap<&str, i32> =
        HashMap::from([("A", 1), ("X", 1), ("B", 2), ("Y", 2), ("C", 3), ("Z", 3)]);

    INPUT
        .lines()
        .map(|line| {
            line.split(" ")
                .map(|value| *to_int.get(value).unwrap())
                .collect::<Vec<i32>>()
        })
        .collect::<Vec<Vec<i32>>>()
}

fn points(them: i32, me: i32) -> i32 {
    if me == them {
        3 + me
    } else if me == WIN[(them - 1) as usize] {
        6 + me
    } else {
        0 + me
    }
}

fn play(them: i32, outcome: i32) -> i32 {
    if outcome == 2 {
        them
    } else if outcome == 1 {
        LOSE[(them - 1) as usize]
    } else {
        WIN[(them - 1) as usize]
    }
}

#[test]
fn solve() {
    let p: i32 = input()
        .iter()
        .map(|round| points(round[0], round[1]))
        .collect::<Vec<i32>>()
        .iter()
        .sum();
    println!("Part 1: {:?}", p);

    let p: i32 = input()
        .iter()
        .map(|round| points(round[0], play(round[0], round[1])))
        .collect::<Vec<i32>>()
        .iter()
        .sum();
    println!("Part 2: {:?}", p);
}
